### PR TITLE
Add Type System Printer

### PR DESCRIPTION
**Pull request by schrockn**
This adds a printer to translate the result of an introspection query
to the grammar that is currently used as shorthand in the specification.
This should be treated as the de facto specification until a formalized
grammar is produced.

The next step is to do the transformation in the other
direction.

A tool that should be built on top of this is one that executes
the introspection query against an arbitrary GraphQL HTTP endpoint
(specified per the upcoming graphql-http spec) and then performs
acceptable tests with that tool.

In the interim we hope that this will be useful for implementors
building and debugging their own type systems.

### CODE REVIEW

File: src/type/__tests__/introspection.js [hunk 1]
```diff
@@ -36,110 +36,12 @@ describe('Introspection', () => {
       await graphql(EmptySchema, introspectionQuery)
     ).to.deep.equal({
       data: {
-        schemaType: {
```
Comment (leebyron) [inline src/type/__tests__/introspection.js]: Is this an unrelated change?

File: src/type/printer.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  ENUM: 'ENUM',
+  INPUT_OBJECT: 'INPUT_OBJECT',
+  LIST: 'LIST',
+  NON_NULL: 'NON_NULL',
+};
+
+// e.g Int, Int!, [Int!]!
+function printTypeDecl(type) {
+  if (type.ofType === undefined || type.ofType === null) {
+    return type.name;
```
Comment (dschafer) [inline src/type/printer.js]: Why not just do this at the end instead of the Error?

File: src/type/printer.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+function printArgs(field) {
+  if (field.args.length === 0) {
+    return '';
+  }
+  return '(' + field.args.map(printArg).join(', ') + ')';
+}
+
+function printFields(fields) {
+  return fields.map(f => '  ' + printField(f)).join('\n');
+  // var strs = [];
```
Comment (dschafer) [inline src/type/printer.js]: rm

File: src/type/definition.js [hunk 4]
```diff
@@ -102,6 +102,14 @@ export function isCompositeType(type: ?GraphQLType): boolean {
   );
 }
 
+export function typeHasFields(type: ?GraphQLType): boolean {
```
Comment (leebyron) [inline src/type/definition.js]: this already exists as `isCompositeType` in this file

Comment (schrockn) [inline src/type/definition.js]: That has different semantics. It does not include input types and includes unions

File: src/type/__tests__/introspection.js (hunk 1 above)
Comment (schrockn) [inline src/type/__tests__/introspection.js]: No. The Pre-existing introspection query was incorrect.

File: src/type/introspection.js [hunk 5]
```diff
@@ -108,8 +109,7 @@ var __Type = new GraphQLObjectType({
         includeDeprecated: { type: GraphQLBoolean, defaultValue: false }
       },
       resolve(type, {includeDeprecated}) {
-        if (type instanceof GraphQLObjectType ||
-            type instanceof GraphQLInterfaceType) {
+        if (typeHasFields(type)) {
```
Comment (leebyron) [inline src/type/introspection.js]: This is not a safe change. Unions do not define any fields, I don't think it defines `getFields()`

File: src/type/printer.js (hunk 3 above)
Comment (schrockn) [inline src/type/printer.js]: √

File: src/type/printer.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
+/**
+ *  Copyright (c) 2015, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+var TypeKind = {
```
Comment (dschafer) [inline src/type/printer.js]: This feels weird to me, not sure why.

File: src/type/printer.js (hunk 2 above)
Comment (schrockn) [inline src/type/printer.js]: Not following

File: src/type/introspection.js (hunk 5 above)
Comment (schrockn) [inline src/type/introspection.js]: Look at the implementation of typeHasFields. It does not include unions

File: src/type/printer.js (hunk 6 above)
Comment (schrockn) [inline src/type/printer.js]: more specific?

File: src/type/definition.js [hunk 7]
```diff
@@ -102,6 +102,14 @@ export function isCompositeType(type: ?GraphQLType): boolean {
...
+  return (
+    type instanceof GraphQLObjectType ||
+    type instanceof GraphQLInterfaceType ||
+    type instanceof GraphQLInputObjectType
```
Comment (dschafer) [inline src/type/definition.js]: This is a change in behavior; `GraphQLInputObjectType` shouldn't have `fields`, it should have `inputFields`, right?

File: src/type/definition.js (hunk 4 above)
Comment (leebyron) [inline src/type/definition.js]: Ah, my mistake

File: src/type/introspection.js (hunk 5 above)
Comment (leebyron) [inline src/type/introspection.js]: Misread the previous, but this is still probably not safe as now InputObject is going to present fields as the wrong type and in the wrong place in introspection

File: src/type/printer.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  UNION: 'UNION',
+  ENUM: 'ENUM',
+  INPUT_OBJECT: 'INPUT_OBJECT',
+  LIST: 'LIST',
+  NON_NULL: 'NON_NULL',
+};
+
+// e.g Int, Int!, [Int!]!
+function printTypeDecl(type) {
+  if (type.ofType === undefined || type.ofType === null) {
```
Comment (leebyron) [inline src/type/printer.js]: I recommend doing instanceof checks instead of directly checking the value of `ofType` - we dont use this pattern elsewhere

File: src/type/introspection.js (hunk 5 above)
Comment (schrockn) [inline src/type/introspection.js]: Yup I was wrong on that count. Will fix. Forgot about inputFields

File: src/type/definition.js (hunk 7 above)
Comment (schrockn) [inline src/type/definition.js]: yup. will fix.

File: src/type/printer.js (hunk 8 above)
Comment (leebyron) [inline src/type/printer.js]: sorry not instanceof - Kind checks.


```
function printTypeDecl(type) {
  if (type.kind === TypeKind.LIST) {
    return '[' + printTypeDecl(type.ofType) + ']';
  } else if (type.kind === TypeKind.NON_NULL) {
    return printTypeDecl(type.ofType) + '!';
  }
  return type.name;
}
```

Comment (schrockn) [inline src/type/printer.js]: √

File: src/type/printer.js (hunk 2 above)
Comment (dschafer) [inline src/type/printer.js]:
```
function printTypeDecl(type) {
  if (type.kind === TypeKind.LIST) {
    return '[' + printTypeDecl(type.ofType) + ']';
  } else if (type.kind === TypeKind.NON_NULL) {
    return printTypeDecl(type.ofType) + '!';
  }
  return type.name;
}
```

File: src/type/printer.js (hunk 6 above)
Comment (dschafer) [inline src/type/printer.js]: I guess it feels weird to define this struct only to use every value once, compared to just using the strings inline, or reusing an existing struct if it exists.

File: src/type/printer.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+function printFields(fields) {
+  return fields.map(f => '  ' + printField(f)).join('\n');
+  // var strs = [];
+  // fields.forEach(field => {
+  //   strs.push('  ' + printField(field));
+  // });
+  // return strs.join('\n');
+}
+
+function printIfaceList(type) {
```
Comment (leebyron) [inline src/type/printer.js]: I would prefer spelling out, perhaps `printInterfaces`

File: src/type/printer.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+  return type.name.startsWith('__');
+}
+
+function isBuiltIn(type) {
+  var name = type.name;
+  if (isIntrospectionType(type)) {
+    return true;
+  }
+
+  return name === 'String' || name === 'Boolean' || name === 'Int';
```
Comment (leebyron) [inline src/type/printer.js]: Float?

Comment (leebyron) [inline src/type/printer.js]: ID?

File: src/type/printer.js (hunk 2 above)
Comment (schrockn) [inline src/type/printer.js]: lolz

File: src/type/printer.js (hunk 6 above)
Comment (schrockn) [inline src/type/printer.js]: there's no existing struct. we should probably move this to a common locale

File: src/type/printer.js (hunk 9 above)
Comment (schrockn) [inline src/type/printer.js]: going to go with printInterfaceUsages to distinguish from printing the definition

File: src/type/printer.js (hunk 10 above)
Comment (schrockn) [inline src/type/printer.js]: Da

Comment (leebyron) [inline src/type/printer.js]: I actually think I prefer not excluding these at all - for the same reason we don't exclude them from introspection.

File: src/type/printer.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+  return values.map(v => '  ' + v.name).join('\n');
+}
+
+function printEnum(type) {
+  return `enum ${type.name} {
+${printEnumValues(type.enumValues)}
+}`;
+}
+
+function printType(type) {
```
Comment (leebyron) [inline src/type/printer.js]: Should this also be exported? It seems useful to print only one type in the same way it's useful to introspect on only one type with `__type(name:)`

**Comment by leebyron** [comment]
This is so awesome. Two high-level points of feedback:

1) I'd recommend splitting this into two PRs. One that address any changes or bug fixes to the introspection system and type definition API itself. It was a little hard for me to follow what additions were necessary for type printing and what additions were ancillary.

2) I think we should reflect on the grammar for types and schema a little more:

* It's a bit odd that every type starts with a keyword for the kind of type it is, expect object types. Now that I'm seeing this all come together, I think we should consider the leading keyword for object types be `object`.

```diff
interface Type { ... }
enum Enum { ... }
- type Type { ... }
+ object Type { ... }
```

* We need some way to describe the initial types for query and mutation. I don't have any immediately useful suggestions here, but I think this will be a point of confusion in printed schema.

* Scalars are a little weird in that they don't have definitions, just declarations. I suppose we could expand on this later.

File: src/type/printer.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    case TypeKind.INTERFACE:
+      return printInterface(type);
+    case TypeKind.INPUT_OBJECT:
+      return printInputObject(type);
+    case TypeKind.SCALAR:
+      return printScalar(type);
+    case TypeKind.ENUM:
+      return printEnum(type);
+    default:
+      throw new Error('Invalid kind: ' + type.kind);
```
Comment (leebyron) [inline src/type/printer.js]: Maybe add a test illustrating the reason for this?

File: src/type/printer.js (hunk 10 above)
Comment (schrockn) [inline src/type/printer.js]: The only thing I find strange is how they are only included when you use them. I do not feel strongly however.

File: src/type/printer.js (hunk 11 above)
Comment (schrockn) [inline src/type/printer.js]: not following this comment

File: src/type/printer.js (hunk 12 above)
Comment (schrockn) [inline src/type/printer.js]: reason for what?
